OpenRA.Mods.Common/Warheads/SpawnActorWarhead.cs

```diff
@@ -0,0 +1,221 @@
+		public readonly string Owner;
+
+		[Desc("Should this actor link to the actor which created it?")]
+		public readonly bool LinkToParent = false;
+
+		[Desc("Should actors always be spawned on the ground?")]
+		public readonly bool ForceGround = false;
+
+		[Desc("Should actors spawn facing the same direction as the projectile?")]
+		public readonly bool SetFacing = false;
+
+		[Desc("Don't spawn actors on this terrain.")]
+		public readonly HashSet<string> InvalidTerrain = new();
+
+		[Desc("Defines the image of an optional animation played at the spawning location.")]
+		public readonly string Image;
+
+		[SequenceReference(nameof(Image), allowNullImage: true)]
+		[Desc("Defines the sequence of an optional animation played at the spawning location.")]
+		public readonly string Sequence = "idle";
+
+		[PaletteReference]
+		[Desc("Defines the palette of an optional animation played at the spawning location.")]
+		public readonly string Palette = "effect";
+
+		public readonly bool UsePlayerPalette = false;
+
+		[Desc("List of sounds that can be played at the spawning location.")]
+		public readonly string[] Sounds = Array.Empty<string>();
+
+		void IRulesetLoaded<WeaponInfo>.RulesetLoaded(Ruleset rules, WeaponInfo info)
+		{
+			foreach (var a in Actors)
+			{
+				var actorInfo = rules.Actors[a.ToLowerInvariant()];
+				var buildingInfo = actorInfo.TraitInfoOrDefault<BuildingInfo>();
+
+				if (buildingInfo != null)
+					throw new YamlException($"SpawnActorWarhead cannot be used to spawn building actor '{a}'!");
+			}
+		}
+
+		public override void DoImpact(in Target target, WarheadArgs args)
+		{
+			var firedBy = args.SourceActor;
+			if (!target.IsValidFor(firedBy))
+				return;
+
+			var world = firedBy.World;
+			var map = world.Map;
+			var pos = target.CenterPosition;
+			var targetCell = map.CellContaining(pos);
+			if (!map.Contains(targetCell))
+				return;
+
+			var directHit = false;
+			foreach (var victim in world.FindActorsOnCircle(pos, WDist.Zero))
+			{
+				if (!IsValidAgainst(victim, firedBy))
+					continue;
+
+				if (!victim.Info.HasTraitInfo<IHealthInfo>())
+					continue;
+
+				if (victim.TraitsImplementing<HitShape>()
+					.Any(i => !i.IsTraitDisabled && i.DistanceFromEdge(victim, pos).Length <= 0))
+				{
+					directHit = true;
+					break;
+				}
+			}
+
+			if (!directHit && !IsValidTarget(map.GetTerrainInfo(targetCell).TargetTypes))
+				return;
+
+			foreach (var a in Actors)
+			{
+				var td = new TypeDictionary();
+				var actorName = a.ToLowerInvariant();
+				var ai = map.Rules.Actors[actorName];
+
+				var owner = Owner == null ? firedBy.Owner : world.Players.First(p => p.InternalName == Owner);
+				td.Add(new OwnerInit(owner));
+
+				if (LinkToParent)
+					td.Add(new ParentActorInit(firedBy));
+
+				if (SetFacing)
+					td.Add(new FacingInit(args.ImpactOrientation.Yaw));
+
+				var cachedTarget = target;
+				world.AddFrameEndTask(w =>
```

Comment:
Trivial. But coulf potentially become a long running frame end task.

Reply:
is that a problem? Everything that handles adding or removing actors to world needs to be in frame end task to avoid a System.InvalidOperationException

Reply:
I was thinking of #20903.

But yeah, it may not matter if it's executed as one or multiple tasks, needs to get executed.